keps/sig-node/2535-ensure-secret-pulled-images/README.md

```diff
@@ -165,57 +201,289 @@ ensure all tenants have rights to the images that are already present on a host.
-It is not expected that this will be a significant number of bytes. Will be verified by actual use in Alpha and subsequent metrics in Beta.
+  2. The image is already present on the node and it's either allow-listed or a record
+of its pull with the given credentials is present in the cache:
+
+```mermaid
+sequenceDiagram
+    participant A as Pod A
+    participant kubelet
+    participant CRI
+    participant cache as persistent<br/>cache
+
+    A->>kubelet: request ("img:tag", secret A)
+    kubelet->>CRI: request ID("img:tag")
+    CRI->>kubelet: ID("img:tag")
+    rect rgb(253, 253, 190)
+    alt image present and set not to reverify
+        kubelet->>kubelet: don't reverify
+    else image present in cache
+        kubelet->>cache: GET _pulled_ (ID("img:tag"), secret A)
+        cache->>kubelet: <previous successful auth info>
+    end
+    end
+    kubelet->>A: continue starting pod
+```
 
-See `/var/lib/kubelet/image_manager_state` in [kubernetes/kubernetes#114847](https://github.com/kubernetes/kubernetes/pull/114847)
+##### Failure modes:
+
+We should always fail safe. If there's an error reading from the disk, the record
+MUST be assumed as non-existent. This might cause an addition request to pull an image.
+If it is successful, the original failing record should be overridden to reflect the
+new success.
+
+#### Cache Directory Structure
+
+A new subdirectory is introduced to the kubelet's directory: `image_manager`.
+The directory contains two further subdirectories: `pulling` and `pulled`. The former
+serves for kubelet to store the state about images that it intends to pull. The
+latter subdirectory should contain information about all the images that were
+successfully pulled by the kubelet.
+
+The figure below shows the Go structures that should appear in files in the
+subdirectories described above. These structures **are not part of the REST API**
+but belong to the `imagemanager.kubelet.config.k8s.io` configuration API group.
+They only ever appear as JSON structures in files outside the kubelet memory.
+
+`ImagePullIntent` is the structure that is present in files in the `image_manager/pulling`
+directory. The filename of each such file is a SHA-256 hash of the image spec prepended
+by `sha256~`.
+
+`ImagePulledRecord` is the structure that appears in files in the `image_manager/pulled`
+directory. The filename of each such file is a sanitized image ID provided by the
+container runtime. This KEP expects the image IDs provided to be in the form of `<hash-alg>:<hash-val>`,
+thus the sanitization consists of replacing the `<hash-alg>:` with `<hash-alg>~`.
+
+```go
+type ImagePullIntent struct {
+    metav1.TypeMeta
+
+    ImageSpec string `json:"imageSpec"` // the filename is a hash of this value
+    Credentials ImagePullCredentials `json:"credentials"`
+}
+
+type ImagePulledRecord struct {
+    metav1.TypeMeta
+
+    ImageID string `json:"imageId"` // the filename is a filesystem sanitized image of this value
```

Comment:
is there a clearer name to use for this?

Does this correspond to the image reference returned from CRI? If so, maybe call it image reference and document that connection?

```
	// PullImage pulls an image from the network to local storage using the supplied
	// secrets if necessary. It returns a reference (digest or ID) to the pulled image.
	PullImage(ctx context.Context, image ImageSpec, pullSecrets []v1.Secret, podSandboxConfig *runtimeapi.PodSandboxConfig) (string, error)
	// GetImageRef gets the reference (digest or ID) of the image which has already been in
	// the local storage. It returns ("", nil) if the image isn't in the local storage.
	GetImageRef(ctx context.Context, image ImageSpec) (string, error)
```

Is CRI *required* to return the digest from PullImage / GetImageRef, or does it just *happen* to? I'm wondering how file-safe we can assume the image id is in terms of sanitizing it for a filename.

Reply:
Image reference would likely be a better name for this field indeed.

I believe the CRI returns its own identification of the pulled image from `PullImage()`, which can either be directly the image repository digest, or the CRI internal ID of the image. I am no expert on container runtimes though so I'd rather have @haircommander or @mikebrow confirm my assumption.

Reply:
yeah it's left this way historically, but ends up being an ImageRef. What further complicates the situation is containerd and cri-o do different things with the result of PullImage (containerd returns a digest, cri-o returns a repo+digest). Thus coming from cri-o I don't think we can assume it's file safe